torch/optim/_multi_tensor/adam.py

```diff
@@ -115,13 +126,22 @@ def step(self, closure=None):
             #
             # Decay the first and second moment running average coefficient
             #
+            is_complex = torch.is_complex(params_with_grad[0])
+            if is_complex:
+                grads = self.foreach_view_as_real(grads)
+                exp_avg = self.foreach_view_as_real(exp_avg)
+                exp_avg_sq = self.foreach_view_as_real(exp_avg_sq)
+                params_with_grad = self.foreach_view_as_real(params_with_grad)
+
             torch._foreach_mul_(exp_avg, beta1)
             torch._foreach_add_(exp_avg, grads, alpha=1 - beta1)
 
             torch._foreach_mul_(exp_avg_sq, beta2)
             torch._foreach_addcmul_(exp_avg_sq, grads, grads, 1 - beta2)
 
             if amsgrad:
+                if is_complex:
+                        max_exp_avg_sq = self.foreach_view_as_real(max_exp_avg_sq)
```

Comment:
How is this one updated? I don't see it being modified inplace?

Reply:
Do you mean updating `max_epx_avg_sq` in the `state['max_exp_avg_sq']`?

Reply:
max_exp_avg_sq comes from that state and we should modify it inplace so that the state is updated by side effect.

Reply:
Ok, then I try to explain here, the `max_exp_avg_sq` is complex whose real part & imaginary parts correspond to the `max_exp_avg_sq` of the real parts and imaginary parts of parameters respectively, so we may need to turn the `max_exp_avg_sq` to be viewed as real numbers by `torch.view_as_real()`. I don't think `max_exp_avg_sq` is same for the real parts and imaginary parts of parameters.